### PR TITLE
Fix torch build on ROCm by avoiding dynamically loaded cuCtxGetDevice

### PR DESCRIPTION
**Pull request by will-HPE**
Macro testing on `__HIP_PLATFORM_HCC__` as workaround ROCm's not supporting PFN_cuCtxGetDevice (does support cuCtxGetDevice -> hipCtxGetDevice).

## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).
https://github.com/horovod/horovod/issues/3927
## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/GOVERNANCE.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by will-HPE** [comment]
Sorry, I keep using ctrl-enter as enter and that submits on here.
So the submission was premature.

**Comment by paklui** [comment]
Hi @will-HPE , small comment, I think `__HIP_PLATFORM_HCC__` may not get defined for ROCm, I think starting ~ ROCm 5.x? it uses `__HIP_PLATFORM_AMD__` instead of `__HIP_PLATFORM_HCC__`. https://docs.amd.com/bundle/HIP-Programming-Guide-v5.1/page/Transitioning_from_CUDA_to_HIP.html#d3937e736 . Do you think about below? Thanks

```
$diff ~/cuda_util.cc.modified ~/cuda_util.cc.will.unmod
32c32,33
< #if HAVE_CUDA
---
> #ifndef __HIP_PLATFORM_HCC__
> #if HAVE_GPU
51a53
> #endif
58c60
< #if HAVE_CUDA
---
> #ifndef __HIP_PLATFORM_HCC__
```

**Comment by paklui** [comment]
I tired the patch and it appears to work

File: horovod/torch/cuda_util.cc
Comment (maxhgerlach) [inline horovod/torch/cuda_util.cc:33]:
```suggestion
#if !HAVE_ROCM
```

File: horovod/torch/cuda_util.cc
Comment (maxhgerlach) [inline horovod/torch/cuda_util.cc:53]:
```suggestion
#endif
```

File: horovod/torch/cuda_util.cc
Comment (maxhgerlach) [inline horovod/torch/cuda_util.cc:60]:
```suggestion
#if !HAVE_ROCM
```

**Review by maxhgerlach — CHANGES_REQUESTED** [review]
Thank you for providing this fix, @will-HPE, and thank you for testing it on your system, @paklui!

Please note that you would need to sign off your commit for DCO (see https://github.com/horovod/horovod/pull/3928/checks?check_run_id=13614457547) before we can merge it.

One other note: In the Horovod codebase we typically use macros `HAVE_CUDA` and `HAVE_ROCM` that are set in the build process. Would it be possible to use these rather than `__HIP_PLATFORM_HCC__`? (See my suggested code changes.)

**Comment by will-HPE** [comment]
Max,

I think I've completed the requested changes. But, I may have done them sub-optimally. Could you take another look and see if everything is as it should be?

**Review by maxhgerlach — APPROVED** [review]

**Comment by maxhgerlach** [comment]
> I think I've completed the requested changes. But, I may have done them sub-optimally. Could you take another look and see if everything is as it should be?

Thanks for the update including the sign-off line on the commit, @will-HPE! Apparently GitHub didn't include one with the suggested edits, so I overrode the DCO check manually. I think in this case the attribution should be clear.

I've triggered the Horovod CI pipeline, which will build and test on Nvidia GPUs, happy to merge if it asses. For confirming that the build works on AMD GPUs we rely on user input, unfortunately.

**Pull request merged**
